third_party/amd/lib/TritonAMDGPUToLLVM/BuiltinFuncToLLVM.cpp

```diff
@@ -98,6 +111,44 @@ class CallOpConversion : public mlir::RewritePattern {
     rewriter.replaceOp(callOp, loadVal);
     return mlir::success();
   }
+
+  LogicalResult convertToLLVMIntrinsic(LLVM::CallOp callOp,
+                                       mlir::PatternRewriter &rewriter) const {
+    StringRef calleeName = callOp.getCallee().value();
+
+    auto operands = callOp.getOperands();
+    auto result = callOp.getResult();
+
+    LLVM::LLVMFunctionType calleeType = callOp.getCalleeType().value();
+    Type returnType = calleeType.getReturnType();
+
+    LLVM::FastmathFlagsAttr flags;
+    auto loc = callOp.getLoc();
+
+    Operation *replacementOp = nullptr;
+    if (calleeName == "__ocml_trition_abs") {
+      assert(operands.size() == 1);
+      if (type::isInt(returnType)) {
+        replacementOp = rewriter.create<LLVM::AbsOp>(
+            loc, returnType, operands[0], /*is_int_min_poison=*/false);
+      } else if (type::isFloat(returnType)) {
+        replacementOp =
+            rewriter.create<LLVM::FAbsOp>(loc, returnType, operands[0], flags);
+      }
+    } else if (calleeName == "__ocml_trition_llrint") {
+      assert(operands.size() == 1);
+      // Note, LrintOp and LlrintOp result in a code-gen error
```

Comment:
Is this the op sequence we see in cuda2gcn.bc? Do you know what's the reason?

Reply:
Yes, here are the definitions of `llrint` from the disassembled `cuda2gcn.bc`

```llvm
; Function Attrs: convergent mustprogress nofree norecurse nounwind willreturn memory(none)
define linkonce_odr protected i64 @__nv_llrint(double noundef %0) local_unnamed_addr #2 {
  %2 = tail call double @__ocml_rint_f64(double noundef %0) #12
  %3 = fptosi double %2 to i64
  ret i64 %3
}

; Function Attrs: convergent mustprogress nofree norecurse nounwind willreturn memory(none)
define linkonce_odr protected i64 @__nv_llrintf(float noundef %0) local_unnamed_addr #2 {
  %2 = tail call float @__ocml_rint_f32(float noundef %0) #12
  %3 = fptosi float %2 to i64
  ret i64 %3
}
```

And here are some definitions from the disassembled `ocml.bc`

```llvm
; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define linkonce_odr hidden noundef double @__ocml_rint_f64(double noundef %0) local_unnamed_addr #0 {
  %2 = tail call double @llvm.rint.f64(double %0)
  ret double %2
}

; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none)
define linkonce_odr hidden noundef float @__ocml_rint_f32(float noundef %0) local_unnamed_addr #0 {
  %2 = tail call float @llvm.rint.f32(float %0)
  ret float %2
}
```

Reply:
Generally speaking, 

```cpp
replacementOp = rewriter.create<LLVM::LlrintOp>(loc, returnType, operands[0]);
```
should be enough. But, the llvm crashes during instruction selection regarding the AMDGPU backend - i.e.,

```llvm
  %22 = getelementptr float, ptr addrspace(1) %19, i64 %21, !dbg !13
  %23 = getelementptr float, ptr addrspace(1) %18, i64 %21, !dbg !14
  %24 = load float, ptr addrspace(1) %23, align 4, !dbg !12
  %25 = load float, ptr addrspace(1) %22, align 4, !dbg !15
  %26 = fadd float %25, %24, !dbg !16
  %27 = tail call i64 @llvm.llrint.i64.f32(float %26), !dbg !17
  %28 = sitofp i64 %27 to float, !dbg !18
  %29 = getelementptr float, ptr addrspace(1) %20, i64 %21, !dbg !19
  store float %28, ptr addrspace(1) %29, align 4, !dbg !18
  br label %.thread1, !dbg !18, !amdgpu.uniform !6

.thread1:                                         ; preds = %16, %4
  call void @llvm.amdgcn.end.cf.i64(i64 %15), !dbg !20
  ret void, !dbg !20
}
LLVM ERROR: unsupported libcall legalization
```
```bash                                                                                                                                                                                                             
Thread 1 "python3" received signal SIGABRT, Aborted.                                                                                                                                                          
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350506304) at ./nptl/pthread_kill.c:44                                                                                                       
44      ./nptl/pthread_kill.c: No such file or directory.                                                                                                                                                     
(gdb) where                                                                                                                                                                                                   
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140737350506304) at ./nptl/pthread_kill.c:44                                                                                                   
#1  __pthread_kill_internal (signo=6, threadid=140737350506304) at ./nptl/pthread_kill.c:78                                                                                                                   
#2  __GI___pthread_kill (threadid=140737350506304, signo=signo@entry=6) at ./nptl/pthread_kill.c:89                                                                                                           
#3  0x00007ffff7cce476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26                                                                                                                         
#4  0x00007ffff7cb47f3 in __GI_abort () at ./stdlib/abort.c:79                                                                                                                                                
#5  0x00007ffe1dca8e00 in llvm::report_fatal_error(llvm::Twine const&, bool) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so                                                             
#6  0x00007ffe1dca8c18 in llvm::report_fatal_error(char const*, bool) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so                                                                    
#7  0x00007ffe1adb8e95 in llvm::SITargetLowering::LowerCall(llvm::TargetLowering::CallLoweringInfo&, llvm::SmallVectorImpl<llvm::SDValue>&) const ()                                                          
   from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#8  0x00007ffe1c4cbf80 in llvm::TargetLowering::LowerCallTo(llvm::TargetLowering::CallLoweringInfo&) const () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#9  0x00007ffe1c4a29e9 in (anonymous namespace)::SelectionDAGLegalize::ExpandLibCall(llvm::RTLIB::Libcall, llvm::SDNode*, std::vector<llvm::TargetLoweringBase::ArgListEntry, std::allocator<llvm::TargetLoweringBase::ArgListEntry> >&&, bool) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#10 0x00007ffe1c4a24c1 in (anonymous namespace)::SelectionDAGLegalize::ExpandLibCall(llvm::RTLIB::Libcall, llvm::SDNode*, bool) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#11 0x00007ffe1c4a1e2e in (anonymous namespace)::SelectionDAGLegalize::ExpandFPLibCall(llvm::SDNode*, llvm::RTLIB::Libcall, llvm::SmallVectorImpl<llvm::SDValue>&) ()
   from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#12 0x00007ffe1c493b47 in (anonymous namespace)::SelectionDAGLegalize::ConvertNodeToLibcall(llvm::SDNode*) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#13 0x00007ffe1c47c9da in (anonymous namespace)::SelectionDAGLegalize::LegalizeOp(llvm::SDNode*) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#14 0x00007ffe1c47b626 in llvm::SelectionDAG::Legalize() () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#15 0x00007ffe1c57c5cf in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#16 0x00007ffe1c57a0cf in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#17 0x00007ffe1c576b3f in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#18 0x00007ffe1ab830ae in AMDGPUDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#19 0x00007ffe1c85c7ef in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#20 0x00007ffe1d9eb360 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#21 0x00007ffe1d2b18dc in (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#22 0x00007ffe1d9eba41 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /home/ravil/work/triton/triton/python/triton/_C/libtriton.so
#23 0x00007ffe1a2785ca in translateLLVMIRToASM (module=..., Python Exception <class 'gdb.error'>: There is no member named _M_p.
triple=, Python Exception <class 'gdb.error'>: There is no member named _M_p.
proc=, Python Exception <class 'gdb.error'>: There is no member named _M_p.
features=, flags=std::vector of length 0, capacity 0, enable_fp_fusion=true, isObject=false)
    at /home/ravil/work/triton/triton/python/src/llvm.cc:118
#24 0x00007ffe1a27e4bd in init_triton_llvm(pybind11::module_&&)::$_8::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, bool, bool) const (
    this=0x69259d8, Python Exception <class 'gdb.error'>: There is no member named _M_p.
llvmIR=, Python Exception <class 'gdb.error'>: There is no member named _M_p.
triple=, Python Exception <class 'gdb.error'>: There is no member named _M_p.
proc=, Python Exception <class 'gdb.error'>: There is no member named _M_p.
````

It is something for a follow-up ticket

Reply:
ok. please file an issue to track this internally.

Reply:
@ravil-mobile Once you have a ticket number, please attach it as a comment in this thread.

Reply:
Hi @jerryyin, the reproducible example is ready